### PR TITLE
Added download controls for study view charts

### PR DESCRIPTION
**Pull request by onursumer**
# What? Why?
Partially addressed download functionality for the study view components.

Changes proposed in this pull request:
- Added download icon into `ChartHeader` (related to https://github.com/cBioPortal/cbioportal/issues/4155)
- Added full download functionality for `PieChart` (related to https://github.com/cBioPortal/cbioportal/issues/4156)
- Added data download functionality for `SurvivalChart` (related to https://github.com/cBioPortal/cbioportal/issues/4159)

# Checks
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

### CODE REVIEW

File: src/pages/studyView/StudyViewPage.tsx [hunk 1]
```diff
@@ -105,6 +105,8 @@ export default class StudyViewPage extends React.Component<IStudyViewPageProps,
             filters: [],
             onDeleteChart: this.handlers.onDeleteChart
         };
+
+        // TODO enable data download for bar chart, too
```
Comment (onursumer) [inline src/pages/studyView/StudyViewPage.tsx]: This will be addressed after merging #1395

File: src/pages/studyView/charts/ChartContainer.tsx [hunk 2]
```diff
@@ -83,9 +89,15 @@ export class ChartContainer extends React.Component<IChartContainerProps, {}> {
...
-                let firstLine = this.props.chartMeta.displayName + '\tCount'
-                fileDownload(firstLine + '\n' + this.plot.downloadData(), this.fileName);
-
+                if (this.props.downloadData) {
+                    this.props.downloadData().then(data => {
+                        if (data && data.length > 0) {
+                            fileDownload(data, this.fileName);
+                        }
+                    }).catch(() => {
+                        // TODO this.triggerDownloadError();
```
Comment (zhx828) [inline src/pages/studyView/charts/ChartContainer.tsx]: Soon, in the supporting virtual study branch, there are two ways of throwing errors. Error popup or error banner on the top.

File: src/pages/studyView/StudyViewPage.tsx [hunk 3]
```diff
@@ -148,6 +168,14 @@ export default class StudyViewPage extends React.Component<IStudyViewPageProps,
                 props.onResetSelection = this.handlers.resetGeneFilter;
                 props.selectedGenes=this.store.selectedGenes;
                 props.onGeneSelect=this.store.onCheckGene;
+                props.disableGraphicsDownload = true;
+                props.download = [
+                    {
+                        // TODO implement a proper data download function
+                        initDownload: () => Promise.resolve("NA"),
+                        type: 'TSV'
```
Comment (zhx828) [inline src/pages/studyView/StudyViewPage.tsx]: Let's make the type as a Enum

Comment (onursumer) [reply src/pages/studyView/StudyViewPage.tsx]: There is a type for this:
```type ChartDownloadType = 'TSV' | 'SVG' | 'PDF';```

This should be enough for now.

**Comment by zhx828** [comment]
@onursumer the PDF legend overflows.

http://www.cbioportal.org/rc/newstudy?id=acc_tcga

screen shot 2018-09-04 at 1 55 26 pm

File: src/pages/studyView/StudyViewPageStore.ts [hunk 4]
```diff
@@ -879,6 +879,93 @@ export class StudyViewPageStore {
...
+        }
+    }
+
+    public async getSurvivalDownloadData(chartMeta: ChartMeta)
+    {
+        if (this.survivalData.result)
+        {
+            const data: string[] = [];
+
+            const uniqueClinicalAttributeIds: {[clinicalAttributeId: string]: string} = {};
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts]: use map or reduce for this so can assign directly to the variable.

File: src/shared/lib/svgToPdfDownload.ts [hunk 5]
```diff
@@ -1,32 +1,53 @@
...
 }
 
 export default function (filename:string, svg:Element, servletUrl?: string) {
+    const promise = svgToPdfRequest(svg, servletUrl);
+
+    if (!promise) {
+       return false;
+    }
+
+    promise.end((err, res)=>{
```
Comment (alisman) [inline src/shared/lib/svgToPdfDownload.ts]: promise.end.  is that standard?  maybe we use then and catch?

Comment (alisman) [reply src/shared/lib/svgToPdfDownload.ts]: i guess this could be specific to what library returns so ...  just wondering

Comment (onursumer) [reply src/shared/lib/svgToPdfDownload.ts]: @alisman Actually this is not a standard promise. This is a `Request` which extends `Promise`. This code segment was already there, I just needed to make some changes to the existing method to be able to reuse it as a regular promise in the download controls without breaking the previous use-case. I think I should rename the variable here to `request` to make it clear.

**Comment by alisman** [comment]
k no prob

**Pull request merged**